fiber: make the concurrent fiber_join safer (#9311)
tarantool/tarantool: changelogs/unreleased/gh-7562-fiber_join.md

## bugfix/core

> * Fixed a possible inconsistent state entering if the `fiber_join` function is
>   used in a wrong way: now the `fiber_set_joinable` function only works if the
>   fiber is alive and is not joined yet and the `fiber_join` diagnoses a double
>   join as an error if it's possible to detect it (gh-7562).

Reviewer: ```suggestion
* Fixed a possible inconsistent state entering if the `fiber_join` function was
  used incorrectly. Now the `fiber_set_joinable` function only works if the
  fiber is alive and is not joined yet. The `fiber_join` function now treats a double join as an error if it is possible to detect it (gh-7562).
```
